### PR TITLE
Adds login

### PR DESCRIPTION
**Pull request by ZooeyMiller**
- [x] set up route for redirect for user to log in 
- [x] use the github callback
- [x] upsert user in database 
- [x] set session with user information
- [x] create plug for requiring auth
- [ ] protect routes so that only logged in users can view them (no routes to protect right now, will document how to do so when needed though.)
- [x] create plug for requiring user to be an admin
- [x] protect routes so that only admins can view them
- [x] work out how to get actual organisation info for all users
- [x] 100% coverage
- [x] Documentation

#32

### CODE REVIEW

File: lib/library_web/controllers/login_controller.ex [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  """
+  def login(conn, _params) do
+
+    case @elixir_auth_github.login_url_with_scope(["user:email", "read:org"]) do
+      {:err, reason} ->
+        # TODO: set internal server error
+        conn = conn
+              |> put_flash(:error, "Something went wrong on our end, try again later!")
+
+        render conn, "index.html"
```
Comment (katbow) [inline lib/library_web/controllers/login_controller.ex]: Why don't we just pipe the conn through here?
As in
```elixir
conn
|> put_flash(:error, "Something went wrong on our end, try again later!")
|> render("index.html")
```

File: lib/library_web/controllers/login_controller.ex [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  @doc """
+    Creates a github URL with the scope for getting the user's email and orgs
+    renders the index.html if there is a failure constructing it (can only
+    happen when elixir_auth_github is not configured correctly/env variables
+    are not set), and redirects to github login if the url is correct
+  """
+  def login(conn, _params) do
+
+    case @elixir_auth_github.login_url_with_scope(["user:email", "read:org"]) do
+      {:err, reason} ->
```
Comment (katbow) [inline lib/library_web/controllers/login_controller.ex]: If we aren't going to use the reason somewhere below here, we can put it as `_reason` so we don't have unused variable warnings.

File: lib/library_web/controllers/login_controller.ex [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            {:error, "not a dwyl member"} ->
+              conn = conn
+                     |> put_flash(:error, "Sorry, the library can only be used by members of dwyl!")
+
+              redirect(conn, to: page_path(conn, :index))
+            error ->
+              conn = conn
+                     |> put_flash(:error, "Problem adding user, sorry, try again later!")
+
+              redirect(conn, to: page_path(conn, :index))
```
Comment (katbow) [inline lib/library_web/controllers/login_controller.ex]: Same with these ones as the comment above. Remove the `conn=` and just pipe it through to the redirects/renders.

File: lib/library_web/controllers/login_controller.ex [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    case @elixir_auth_github.github_auth(code) do
+      {:error, error} ->
+        conn
+        |> put_flash(:error, "Problem getting info from github, try again later")
+        |> render("index.html")
+      {:ok, user} ->
+        case add_user(user) do
+            {:ok, user} ->
+              conn = conn
+                    |> put_flash(:info, "added to db succesfully")
```
Comment (katbow) [inline lib/library_web/controllers/login_controller.ex]: Can we keep the same capitalisation through the app for flash messages? Most(all?) of the others have the first letter capitalised.

File: lib/library_web/controllers/plugs/require_admin.ex [hunk 5]
```diff
@@ -0,0 +1,23 @@
+defmodule LibraryWeb.Plugs.RequireAdmin do
+  import Plug.Conn
+  import Phoenix.Controller
+  alias LibraryWeb.Router.Helpers
+
+  def init(_params) do
+  end
```
Comment (katbow) [inline lib/library_web/controllers/plugs/require_admin.ex]: We can use the `do:` syntax here for one line functions.

File: lib/library_web/controllers/plugs/require_auth.ex [hunk 6]
```diff
@@ -0,0 +1,24 @@
+defmodule LibraryWeb.Plugs.RequireAuth do
+  import Plug.Conn
+  import Phoenix.Controller
+  alias LibraryWeb.Router.Helpers
+
+  def init(_params) do
+  end
```
Comment (katbow) [inline lib/library_web/controllers/plugs/require_auth.ex]: Same here for `do:`

File: lib/library_web/controllers/login_controller.ex (hunk 3 above)
Comment (ZooeyMiller) [reply lib/library_web/controllers/login_controller.ex]: The only reason I didn't pipe here was due to using `conn` in the `page_path` as well as in the `redirect` function. Refactoring in a slightly different way, will see what you think when I push up.

File: lib/library_web/controllers/plugs/require_admin.ex (hunk 5 above)
Comment (ZooeyMiller) [reply lib/library_web/controllers/plugs/require_admin.ex]: Since this function doesn't return anything, the one like `do:` syntax breaks the code.

**Pull request merged**